### PR TITLE
KAFKA-12637: Remove deprecated PartitionAssignor interface

### PR DESCRIPTION
**Pull request by dengziming**
*More detailed description of your change*
1. Remove PartitionAssignor and related classes, update docs
2. Move unit test
3. fix some related typo

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dengziming** [comment]
@ableegoldman , Hi, PTAL, thank you.

File: docs/upgrade.html
Comment (ableegoldman) [inline docs/upgrade.html]:
```suggestion
        <li>The deprecated <code>o.a.k.clients.consumer.internals.PartitionAssignor</code> class has been removed. Please use <code>ConsumerPartitionAssignor</code> instead.</li>
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java]: This particular test doesn't make sense any more, since there is no "old" assignor type now that PartitionAssignor is removed

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java]:
```suggestion
    public void shouldInstantiateAssignor() {
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:52]: Can you add one more test like this one, but with more than one assignor in the list to instantiate?

**Review by ableegoldman — COMMENTED** [review]
Thanks for the PR, few minor suggestions but overall looks good

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (dengziming) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:52]: Add `shouldInstantiateListOfAssignors`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (dengziming) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java]: I changed its name to `shouldInstantiateFromListOfClassTypes` and add a similar `shouldInstantiateFromClassType` to test "new" assignor, WDYT.

File: docs/upgrade.html
Comment (dengziming) [reply docs/upgrade.html]: I reworded this and added a "and related classes" since we also removed `PartitionAssignorAdaptor`, WDYT?

**Review by dengziming — COMMENTED** [review]
Thank you for your reply @ableegoldman , call for a second review 😄.

File: docs/upgrade.html
Comment (ableegoldman) [reply docs/upgrade.html]: I changed it to the singular `class` intentionally since the `PartitionAssignorAdaptor` was never a public API, the only thing that was exposed to users -- and therefore the only thing they care or need to know about -- was the `PartitionAssignor`. Does that make sense?

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignor.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignor.java]: nit: this should be package private I think

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:52]: we should do this `assertTrue` thing for the CooperativeStickyAssignor as well

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: I think it would make sense to style this test (and `shouldInstantiateFromListOfClassTypes` below) more like `shouldInstantiateAssignors` now, ie where we actually validate the assignors that are returned (eg `assertTrue(assignors.get(0) instanceof StickyAssignor)`). Previously this test was just making sure that we adaptor would work and we wouldn't throw an exception when constructing the consumer, that's why it's like this

File: docs/upgrade.html
Comment (ableegoldman) [inline docs/upgrade.html]:
```suggestion
        <li>The deprecated <code>org.apache.kafka.clients.consumer.internals.PartitionAssignor</code> class has been removed. Please use
```

See my reply on the original comment 🙂

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java
Comment (dengziming) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: `assignors ` is not visible in `KafkaConsumer`, we should add a `getAssignors()` in `KafkaConsumer` to test this. Maybe the other tests are enough to verify that the `getAssignorInstances` operate rightly, How about remove these 2 test cases?

Comment (ableegoldman) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: Ah, I was suggesting to just replicate the `shouldInstantiateAssignor` and `shouldInstantiateListOfAssignors` tests exactly, but with the `classTypes` being eg `StickyAssignor.class` instead of `StickyAssignor.class.getName()`. For example

```
classNames = Collections.singletonList(StickyAssignor.class);
List<ConsumerPartitionAssignor> assignors = getAssignorInstances(classNames, Collections.emptyMap());
assertTrue(assignors.get(0) instanceof StickyAssignor);
```

Comment (dengziming) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: It seems that `getAssignorInstances()` can only accept `List<String>` as the first parameter, so this could not work.😳

Comment (ableegoldman) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: Ah, yeah, you'd need to do something more like what actually happens in the actual KafkaConsumer/`getAssignorInstances` code. eg
```
@Test
@SuppressWarnings("unchecked")
public void shouldInstantiateAssignorClass() {
    Object classTypes = Collections.singletonList(StickyAssignor.class);
    List<ConsumerPartitionAssignor> assignors = getAssignorInstances((List<String>) classTypes, Collections.emptyMap());
     assertTrue(assignors.get(0) instanceof StickyAssignor);
}
```

Comment (ableegoldman) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: It's super awkward, obviously, but since this is what happens when we process the configs in the real code we should try to replicate that in the test

Comment (dengziming) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: Yes, this awkward, I fixed this with some code refactoring and fixed the comment in upgrade.html, PTAL, Thank you for your patience.

Comment (ableegoldman) [reply clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignorTest.java:73]: Nice, thanks for the update. Looks good

**Review by ableegoldman — APPROVED** [review]
LGTM, please ping me again when the build completes and I'll merge

**Comment by dengziming** [comment]
ping @ableegoldman .😉

**Comment by ableegoldman** [comment]
Hm, the build was aborted -- looks like the gradle daemon was killed or crashed for some reason? I'll kick off a new run, let's hope this one passes

**Comment by ableegoldman** [comment]
Well there were a large number of failures, but all of them are unrelated. Left a comment/reopened the ticket for all of the following:

kafka.connect.mirror.integration.MirrorConnectorsIntegrationTest.testReplication() 
kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions() 
kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testOneWayReplicationWithAutoOffsetSync()
kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testDeleteConnector kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTest.testOneWayReplicationWithAutoOffsetSync()
kafka.connect.mirror.integration.MirrorConnectorsIntegrationTest.testOneWayReplicationWithAutoOffsetSync()

**Comment by ableegoldman** [comment]
Merged to trunk -- thanks @dengziming !

**Pull request merged**
